src/H5Dmpio.c

```diff
@@ -2954,464 +3398,1569 @@ H5D__chunk_redistribute_shared_chunks(const H5D_io_info_t *io_info, const H5D_ty
+#else
+                base_buf = (MPI_Aint)chunk_list[0].buf;
+#endif
+
+                base_offset = chunk_list[0].fspace_info.chunk_new.offset;
+            }
+
+            for (i = 0, chunk_count = 0; i < num_entries; i++) {
+                if (op_type == H5D_IO_OP_READ) {
+                    /*
+                     * If this is a read operation that is occurring as
+                     * part of a dataset write (read-modify-write cycle),
+                     * don't add the chunk to the list of addresses being
+                     * read if the chunk is being fully overwritten
+                     */
+                    if (!chunk_list[i].need_read)
+                        continue;
+                    else {
+                        /* Otherwise, set our base chunk offset and base chunk
+                         * data read buffer if we haven't already
+                         */
+                        if (!base_read_vars_set) {
+#if MPI_VERSION >= 3
+                            if (MPI_SUCCESS != (mpi_code = MPI_Get_address(chunk_list[i].buf, &base_buf)))
+                                HMPI_GOTO_ERROR(FAIL, "MPI_Get_address failed", mpi_code)
+#else
+                            base_buf = (MPI_Aint)chunk_list[i].buf;
+#endif
+
+                            base_offset = chunk_list[i].fspace_info.chunk_current.offset;
+
+                            base_read_vars_set = TRUE;
+                        }
+                    }
+                }
+
+                /*
+                 * Set the current chunk entry's offset in the file, relative to
+                 * the first chunk entry
+                 */
+                if (op_type == H5D_IO_OP_READ) {
+                    file_offset_array[chunk_count] =
+                        (MPI_Aint)(chunk_list[i].fspace_info.chunk_current.offset - base_offset);
+                }
+                else {
+                    file_offset_array[chunk_count] =
+                        (MPI_Aint)(chunk_list[i].fspace_info.chunk_new.offset - base_offset);
+                }
+
+                /*
+                 * Ensure the chunk list is sorted in ascending ordering of
+                 * offset in the file
+                 */
+                if (chunk_count)
+                    HDassert(file_offset_array[chunk_count] > file_offset_array[chunk_count - 1]);
+
+                /* Set the current chunk entry's size for the I/O operation */
+                if (op_type == H5D_IO_OP_READ) {
+                    H5_CHECK_OVERFLOW(chunk_list[i].fspace_info.chunk_current.length, hsize_t, int);
+                    length_array[chunk_count] = (int)chunk_list[i].fspace_info.chunk_current.length;
+                }
+                else {
+                    H5_CHECK_OVERFLOW(chunk_list[i].fspace_info.chunk_new.length, hsize_t, int);
+                    length_array[chunk_count] = (int)chunk_list[i].fspace_info.chunk_new.length;
```

Comment:
It might be worth setting up a convenience pointer to chunk_current/chunk_new to avoid these read/write branches and make the code a little more readable.

Reply:
I think that's a good idea, especially for readability.

Reply:
Fixed.